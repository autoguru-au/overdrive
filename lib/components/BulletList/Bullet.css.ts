import { style, styleVariants } from '@vanilla-extract/css';

import { overdriveTokens as vars } from '../../themes/theme.css';

export const root = styleVariants({
	default: {
		display: 'grid',
		gridTemplateColumns: 'min-content 1fr',
		gridGap: vars.space['4'],
		marginBottom: vars.space['2'],

		':last-of-type': {
			marginBottom: 0,
		},

		'::before': {
			boxSizing: 'content-box',
			display: 'block',
			alignSelf: 'flex-start',
			width: vars.space['2'],
			height: vars.space['2'],
			marginTop: `calc((1.6em - ${vars.space['2']}) / 2)`,
			content: "' '",
			backgroundColor: vars.typography.colour.dark,
		},
	},
	circle: {
		'::before': {
			borderRadius: '50%',
		},
	},
	disc: {
		'::before': {
			border: `2px solid ${vars.typography.colour.dark}`,
			borderRadius: '50%',
			backgroundColor: 'transparent',
		},
	},
	square: {
		'::before': {
			border: `2px solid ${vars.typography.colour.dark}`,
			backgroundColor: 'transparent',
		},
	},
});

export const noDot = style({
	'::before': {
		visibility: 'hidden',
	},
});
