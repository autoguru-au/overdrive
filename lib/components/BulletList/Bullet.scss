@import '../../theme/tokens/facade';

.root {
	$dotSize: 8px;

	display: grid;
	grid-template-columns: min-content 1fr;
	grid-gap: var(--global--space--4);

	&:not(:last-of-type) {
		margin-bottom: var(--global--space--2);
	}

	&::before {
		display: block;
		align-self: flex-start;
		width: $dotSize;
		height: $dotSize;
		margin-top: calc((1.6em - #{$dotSize}) / 2);
		content: '';
		background-color: getColour(gray, 900);
	}

	&.circle::before {
		border-radius: 50%;
	}

	&.disc::before {
		border: 2px solid getColour(gray, 900);
		border-radius: 50%;
		background-color: transparent;
	}

	&.square::before {
		border: 2px solid getColour(gray, 900);
		background-color: transparent;
	}
}

.noDot {
	&::before {
		visibility: hidden;
	}
}
