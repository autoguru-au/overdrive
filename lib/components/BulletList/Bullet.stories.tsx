import { ComponentMeta, ComponentStory } from '@storybook/react';
import * as React from 'react';
import { ComponentProps } from 'react';

import { Text } from '../Text';

import { Bullet, BulletList } from '.';

export default {
	title: 'Foundation/List/BulletList',
	component: BulletList,
	decorators: [
		(story) => (
			<div style={{ maxWidth: '500px', width: '100%' }}>{story()}</div>
		),
	],
} as ComponentMeta<typeof BulletList>;

const StandardBulletListTemplate: ComponentStory<typeof BulletList> = (
	args,
) => (
	<BulletList {...args}>
		<Bullet>
			<Text>
				For some, it could be argued that air-conditioning is the most
				important feature of a car, excluding all the mechanical bits
				that make it go, stop and turn.
			</Text>
		</Bullet>
		<Bullet>
			<Text>
				There are many components that have to work together to ensure
				the air-conditioning system is working correctly and, with so
				many contributing parts, itâ€™s clear the system should only be
				serviced or repaired by a qualified technician.
			</Text>
		</Bullet>
	</BulletList>
);
const NestedBulletListTemplate: ComponentStory<typeof BulletList> = (args) => (
	<BulletList {...args}>
		<Bullet>
			<Text>
				For some, it could be argued that air-conditioning is the most
				important feature of a car, excluding all the mechanical bits
				that make it go, stop and turn.
			</Text>
		</Bullet>
		<Bullet>
			<BulletList {...args}>
				<Bullet>
					<Text>
						For some, it could be argued that air-conditioning is
						the most important feature of a car, excluding all the
						mechanical bits that make it go, stop and turn.
					</Text>
				</Bullet>
				<Bullet>
					<BulletList {...args}>
						<Bullet>
							<Text>
								For some, it could be argued that
								air-conditioning is the most important feature
								of a car, excluding all the mechanical bits that
								make it go, stop and turn.
							</Text>
						</Bullet>
						<Bullet>
							<BulletList>
								<Bullet>
									<Text>
										For some, it could be argued that
										air-conditioning is the most important
										feature of a car, excluding all the
										mechanical bits that make it go, stop
										and turn.
									</Text>
								</Bullet>
							</BulletList>
						</Bullet>
					</BulletList>
				</Bullet>
			</BulletList>
		</Bullet>
		<Bullet>
			<Text>
				For some, it could be argued that air-conditioning is the most
				important feature of a car, excluding all the mechanical bits
				that make it go, stop and turn.
			</Text>
		</Bullet>
	</BulletList>
);

const standardProps: ComponentProps<typeof BulletList> = {};

export const standard = StandardBulletListTemplate.bind(standardProps);
standard.args = {};

export const nested = NestedBulletListTemplate.bind(standardProps);
nested.args = {};
