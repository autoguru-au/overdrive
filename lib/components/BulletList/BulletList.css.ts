import { style } from '@vanilla-extract/css';

import { themeContractVars as vars } from '../../themes/theme.css';

export const root = style({
	listStyle: 'inside none',
});

export const firstOccurrence = style({
	paddingLeft: vars.space['6'],
});
