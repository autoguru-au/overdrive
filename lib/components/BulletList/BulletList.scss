.root {
	list-style-position: inside;
}

.firstOccurrence {
	padding-left: var(--global--space--6);
}
