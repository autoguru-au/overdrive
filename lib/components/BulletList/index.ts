export { BulletList } from './BulletList';
export { Bullet } from './Bullet';
