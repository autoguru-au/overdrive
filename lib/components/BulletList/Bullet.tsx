import clsx from 'clsx';
import React, {
	FunctionComponent,
	isValidElement,
	ReactChild,
	useContext,
} from 'react';

import styles from './Bullet.scss';
import { BulletList } from './BulletList';
import { BulletListContext, bulletMap } from './context';

interface Props {
	children: ReactChild;
	className?: string;
}

export const Bullet: FunctionComponent<Props> = ({ children, className }) => (
	<li
		className={clsx(
			styles.root,
			bulletMap[useContext(BulletListContext)] ?? '',
			{
				[styles.noDot]:
					isValidElement(children) && children.type === BulletList,
			},
			className,
		)}>
		{children}
	</li>
);
