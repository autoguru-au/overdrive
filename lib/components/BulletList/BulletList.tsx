import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, useContext } from 'react';
import { useStyles } from 'react-treat';

import { Box } from '../Box';
import * as styleRefs from './BulletList.treat';
import { BulletListContext, bulletMap } from './context';

interface Props {
	className?: string;
}

export const BulletList: FunctionComponent<Props> = ({
	children,
	className,
}) => {
	const stack = useContext(BulletListContext);
	const styles = useStyles(styleRefs);

	return (
		<Box
			is="ul"
			className={clsx(
				styles.root,
				{ [styles.firstOccurrence]: stack === -1 },
				className,
			)}>
			<BulletListContext.Provider
				value={stack + 1 >= bulletMap.length ? 0 : stack + 1}>
				{children}
			</BulletListContext.Provider>
		</Box>
	);
};
