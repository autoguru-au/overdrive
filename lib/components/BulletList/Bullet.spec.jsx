import { render } from '@testing-library/react';
import * as React from 'react';
import { Bullet } from './Bullet';
import { BulletList } from './BulletList';

describe('<Bullet />', () => {
	it('should not throw', () => {
		expect(() => render(<Bullet />)).not.toThrow();
	});

	it('should match the snapshot', () => {
		expect(render(<Bullet />).container.firstChild).toMatchSnapshot();
	});

	it.skip('should pass down className', () => {
		expect(
			render(<Bullet className="test-class" />).container.firstChild,
		).toHaveClass('test-class');
	});

	it.skip('should render a dot when child is <BulletList />', () => {
		const { container } = render(
			<Bullet>
				<BulletList>
					<Bullet>test</Bullet>
				</BulletList>
			</Bullet>,
		);

		expect(container.firstChild).toHaveClass('noDot');
	});

	it.skip('should not render a dot when child is not <BulletList />', () => {
		const { container } = render(<Bullet>test</Bullet>);

		expect(container.firstChild).not.toHaveClass('noDot');
	});
});
