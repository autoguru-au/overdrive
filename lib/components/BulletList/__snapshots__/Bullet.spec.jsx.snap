// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Bullet /> > should match the snapshot 1`] = `
<div
  class="theme_OD_Base__j49br80"
  data-od-component="provider"
>
  <li
    class="elementReset_resetVariants__18vhvlid elementReset_resetVariants_as_li__18vhvli11  Bullet_root_default__13ecipx1"
  />
</div>
`;
